_posts/2024-09-09-using-search-latency-monitoring.markdown

```diff
@@ -0,0 +1,193 @@
+ **** **"time_in_millis"****:** **47****,**
+ **** **"current"****:** **0****,**
+ **** **"total"****:** **41**
+ **** **},**
+ **** **"query"****:** **** **{**
+ **** **"time_in_millis"****:** **** **5429****,**
+ **** **"current"****:** **** **13****,**
+ **** **"total"****:** **** **238**
+ **** **},**
+ **** **"fetch"****:** **** **{**
+ **** **"time_in_millis"****:** **** **1230****,**
+ **** **"current"****:** **** **2****,**
+ **** **"total"****:** **** **238**
+ **** **},**
+ **** **"dfs_query"****:** **** **{**
+ **** **"time_in_millis"****:** **** **0****,**
+ **** **"current"****:** **** **0****,**
+ **** **"total"****:** **** **0**
+ **** **},**
+ **** **"expand"****:** **** **{**
+ **** **"time_in_millis"****:** **** **304****,**
+ **** **"current"****:** **1****,**
+ **** **"total"****:** **238**
+ **** **},**
+ **** **"can_match"****:** **** **{**
+ **** **"time_in_millis"****:** **** **0****,**
+ **** **"current"****:** **** **0****,**
+ **** **"total"****:** **** **0**
+ **** **}**
+ **** **}**
+        }
+      }
+    },
+    ...
+  }
+```
+
+### Search phase took time
+
+Introduced in OpenSearch 2.12, the `phase_took` parameter returns phase took times directly in the search response, as shown in the following example:
+
+```json
+GET /my-index-000001/_search?**phase_took**
+{
+  "took" : 105,
+  **"phase_took" : {
+    "dfs_pre_query" : 0,
+    "query" : 69,
+    "fetch" : 22,
+    "dfs_query" : 0,
+    "expand" : 4,
+    "can_match" : 0
+  },**
+  "timed_out" : false,
+  "_shards" : {
+    "total" : 5,
+    "successful" : 5,
+    "skipped" : 0,
+    "failed" : 0
+  },
+  "hits" : {
+    "total" : {
+    "value" : 0,
+    "relation" : "eq"
+  },
+  "max_score" : null,
+  "hits" : [ ]
+}
+```
+
+### Search request slow logs
+
+Introduced in OpenSearch 2.12, search request slow logs provide statistics about search requests that exceed user-defined latency thresholds, as shown in the `opensearch_index_search_slowlog.log` example file:
+
+```json
+[2023-10-30T15:47:42,630][TRACE][c.s.r.slowlog] [runTask-0] 
+took[80.8ms], 
+took_millis[80], 
+phase_took_millis[{expand=0, query=39, fetch=22}], 
+total_hits[4 hits], search_type[QUERY_THEN_FETCH], 
+shards[{total: 10, successful: 10, skipped: 0, failed: 0}], 
+source[{"query":{"match_all":{"boost":1.0}}}], 
+id[]
+```
+
+For more information about how to set up slow log thresholds, see [Search request slow logs](https://opensearch.org/docs/latest/install-and-configure/configuring-opensearch/logs/#search-request-slow-logs).
+
+## Enabling search latency monitoring
+
+To enable these features, visit the [Search settings](https://opensearch.org/docs/latest/install-and-configure/configuring-opensearch/search-settings/) page in the OpenSearch documentation or use the following example API request:
+
+**NOTE**: Search request slow log thresholds can be customized. 
+
+```json
+PUT _cluster/settings
+{
+  "persistent" : {
+    // Coordinator nodes stats
```

Comment:
// Search request level stats enabled (shown as coordinator node stats in node stats api)